<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Integration analysis of multi-sample single cell RNA-seq experiment • SingleCellWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Integration analysis of multi-sample single cell RNA-seq experiment">
<meta property="og:description" content="SingleCellWorkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SingleCellWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/SingleCellWorkshop.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/yunshun/SingleCellWorkshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="SingleCellWorkshop_files/header-attrs-2.10/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Integration analysis of multi-sample single cell RNA-seq experiment</h1>
                        <h4 class="author">Yunshun Chen</h4>
            <address class="author_afil">
      Walter and Eliza Hall Institute of Medical Research, 1G Royal Parade, Parkville, VIC 3052, Melbourne, Australia<br><a class="author_email" href="mailto:#"></a><a href="mailto:yuchen@wehi.edu.au" class="email">yuchen@wehi.edu.au</a>
      </address>
                  
            <h4 class="date">Aug 16, 2021</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/yunshun/SingleCellWorkshop/blob/master/vignettes/SingleCellWorkshop.Rmd"><code>vignettes/SingleCellWorkshop.Rmd</code></a></small>
      <div class="hidden name"><code>SingleCellWorkshop.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Description of your vignette</p>
    </div>
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Single-cell RNA sequencing (scRNA-seq) has become a widely used technique that allows researchers to profile the gene expression and study molecular biology at the cellular level. It provides biological resolution that cannot be achieved with conventional bulk RNA-seq experiments on cell populations.</p>
<p>Here, we provide a detailed workflow for analyzing the 10X single cell RNA-seq data from a single cell RNA expression atlas of human breast tissue <span class="citation">(Pal et al. 2021)</span>. This cell atlas explores the cellular heterogeneity of human mammary gland across different states including normal, pre-neoplastic and cancerous states.</p>
<p>We will be using part of this cell atlas data to demonstrate how to perform a standard analysis for examining one single cell sample, an integration analysis for exploring the cellular heterogeneity across multiple samples, and how to perform differential expression analysis using pseudo-bulk approach. Most of the analysis will be performed using the <em>Seurat</em> package <span class="citation">(Satija et al. 2015)</span>.</p>
</div>
<div id="preliminary" class="section level1">
<h1 class="hasAnchor">
<a href="#preliminary" class="anchor"></a>Preliminary</h1>
<div id="pre-processing-the-raw-data" class="section level2">
<h2 class="hasAnchor">
<a href="#pre-processing-the-raw-data" class="anchor"></a>Pre-processing the raw data</h2>
<p>The raw 10X data come in BCL format. They need to be pre-processed by software tools such as <em>cellranger</em>.</p>
<p>We use <em>cellranger</em> to convert BCL files into the FASTQ files, then align the reads to the human genome, and finally quantify the UMI counts for each gene in each cell. The entire analysis in this workflow is conducted within the R environment using the outputs from <em>cellranger</em>. We do not cover the details of running <em>cellranger</em> as they are beyond the scope of this workflow. The information of how to run <em>cellranger</em> is available at .</p>
</div>
<div id="downloading-the-read-counts" class="section level2">
<h2 class="hasAnchor">
<a href="#downloading-the-read-counts" class="anchor"></a>Downloading the read counts</h2>
<p>In this workshop, we will be using five samples from this published study <span class="citation">(Pal et al. 2021)</span>. These five samples correspond to five individual patients with the following Ids: N1469, N0280, N1105, N0230 and N0123. The accession numbers of the five samples are GSM4909258, GSM4909255, GSM4909260, GSM4909264 and GSM4909267, respectively.</p>
<p>We first create a <code>Data</code> folder under the current working directory. Then we make five subfolders with the patient Ids as the folder names under the <code>Data</code> folder.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"N1469"</span>, <span class="st">"N0280"</span>, <span class="st">"N1105"</span>, <span class="st">"N0230"</span>, <span class="st">"N0123"</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">Samples</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="va">out_dir</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="st">"Data"</span>, <span class="va">Samples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span>
    <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.exists</a></span><span class="op">(</span><span class="va">out_dir</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span><span class="op">(</span><span class="va">out_dir</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>The cellranger output of each sample consists of three data files: a count matrix in <em>mtx.gz</em> format, barcode information in <em>tsv.gz</em> format, and feature information in <em>tsv.gz</em>. We download the three data files from GEO for each of the five samples and store them in each of the subfolders accordingly.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">GSM</span> <span class="op">&lt;-</span> <span class="st">"https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4909nnn/"</span>
<span class="va">url.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">GSM</span>, 
    <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GSM4909258/suppl/GSM4909258_N-NF-Epi-matrix.mtx.gz"</span>,
      <span class="st">"GSM4909255/suppl/GSM4909255_N-N280-Epi-matrix.mtx.gz"</span>, 
      <span class="st">"GSM4909260/suppl/GSM4909260_N-N1105-Epi-matrix.mtx.gz"</span>,
      <span class="st">"GSM4909264/suppl/GSM4909264_N-N1B-Epi-matrix.mtx.gz"</span>,
      <span class="st">"GSM4909267/suppl/GSM4909267_N-MH0023-Epi-matrix.mtx.gz"</span><span class="op">)</span><span class="op">)</span>
<span class="va">url.barcodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"matrix.mtx"</span>, <span class="st">"barcodes.tsv"</span>, <span class="va">url.matrix</span><span class="op">)</span>
<span class="va">url.features</span> <span class="op">&lt;-</span> <span class="st">"https://ftp.ncbi.nlm.nih.gov/geo/series/GSE161nnn/GSE161529/suppl/GSE161529_features.tsv.gz"</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">Samples</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="va">url.matrix</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, destfile<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Data/"</span>, <span class="va">Samples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"/matrix.mtx.gz"</span><span class="op">)</span>, mode<span class="op">=</span><span class="st">"wb"</span><span class="op">)</span> 
    <span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="va">url.barcodes</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, destfile<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Data/"</span>, <span class="va">Samples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"/barcodes.tsv.gz"</span><span class="op">)</span>, mode<span class="op">=</span><span class="st">"wb"</span><span class="op">)</span> 
    <span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="va">url.features</span>, destfile<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Data/"</span>, <span class="va">Samples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"/features.tsv.gz"</span><span class="op">)</span>, mode<span class="op">=</span><span class="st">"wb"</span><span class="op">)</span> 
<span class="op">}</span></code></pre></div>
</div>
</div>
<div id="standard-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#standard-analysis" class="anchor"></a>Standard analysis</h1>
<div id="read-in-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#read-in-the-data" class="anchor"></a>Read in the data</h2>
<p>We load the Seurt package and read in the 10X data for Patient N1469. The object <code>N1469.data</code> is a sparse matrix containing the raw count data of Patient N1469. Rows are features (genes) and columns are cells. By default, the column names of the data are the cell barcodes. To distinguish cells from different patients in the integration analysis later on, the patient Id is added as a prefix to the column names.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span>
<span class="va">N1469.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html">Read10X</a></span><span class="op">(</span>data.dir <span class="op">=</span> <span class="st">"Data/N1469"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">N1469.data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"N1469"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">N1469.data</span><span class="op">)</span>, sep<span class="op">=</span><span class="st">"_"</span><span class="op">)</span></code></pre></div>
<p>We then create a Seurat object <code>N1469</code>. Genes expressed in less than 3 cells are removed. Cells with at least 200 detected genes are kept in the analysis.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">N1469</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts<span class="op">=</span><span class="va">N1469.data</span>, project<span class="op">=</span><span class="st">"N1469"</span>, min.cells<span class="op">=</span><span class="fl">3</span>, min.features<span class="op">=</span><span class="fl">200</span><span class="op">)</span></code></pre></div>
</div>
<div id="quality-control" class="section level2">
<h2 class="hasAnchor">
<a href="#quality-control" class="anchor"></a>Quality control</h2>
<p>Quality control is essential for scRNA-seq analysis. Cells of low quality and genes of low expression shall be removed prior to the analysis.</p>
<p>Two common measures of cell quality are the library size and the number of expressed genes in each cell. The number of unique genes and total molecules (library size) are automatically calculated during <code><a href="https://satijalab.org/seurat/reference/CreateSeuratObject.html">CreateSeuratObject()</a></code>. Another measure to look at is the proportion of reads from mitochondrial genes in each cell. Cells with higher mitochondrial content are more prone to die, and hence should also be from the analysis <span class="citation">(Ilicic et al. 2016)</span>. Here, we calculate the percentages of reads from mitochondrial genes and store them in the metadata of the Seurat object. We use the set of all genes starting with MT- as a set of mitochondrial genes.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">N1469</span><span class="op">[[</span><span class="st">"percent.mt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">N1469</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span></code></pre></div>
<p>The QC metrics of a Seurat object can be viewed as follows.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">N1469</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></code></pre></div>
<pre><code>##                        orig.ident nCount_RNA nFeature_RNA percent.mt
## N1469_AAACATACGGAGCA-1      N1469       7459         1870       2.65
## N1469_AAACATTGCCTCAC-1      N1469       5457         1393       1.89
## N1469_AAACATTGGTAGGG-1      N1469       7547         1964       2.09
## N1469_AAACCGTGCAGTTG-1      N1469        765          328      38.17
## N1469_AAACCGTGCATGAC-1      N1469       3908         1294       5.12
## N1469_AAACGCACTCAGGT-1      N1469       7218         2102       2.13</code></pre>
<p>Scatter plots can be produced for visualizing some of the QC metrics.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeatureScatter.html">FeatureScatter</a></span><span class="op">(</span><span class="va">N1469</span>, feature1 <span class="op">=</span> <span class="st">"nCount_RNA"</span>, feature2 <span class="op">=</span> <span class="st">"percent.mt"</span>, plot.cor<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeatureScatter.html">FeatureScatter</a></span><span class="op">(</span><span class="va">N1469</span>, feature1 <span class="op">=</span> <span class="st">"nCount_RNA"</span>, feature2 <span class="op">=</span> <span class="st">"nFeature_RNA"</span>, plot.cor<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="va">plot1</span> <span class="op">+</span> <span class="va">plot2</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="SingleCellWorkshop_files/figure-html/QC_Scatter_plot-1.png" alt="Scatter plots of QC metrics." width="864"><p class="caption">
Scatter plots of QC metrics.
</p>
</div>
<p>For this particular data, we filter cells that have unique feature less than 500 and with &gt;20% mitochondrial counts.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">N1469</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">N1469</span>, subset <span class="op">=</span> <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">500</span> <span class="op">&amp;</span> <span class="va">percent.mt</span> <span class="op">&lt;</span> <span class="fl">20</span><span class="op">)</span></code></pre></div>
</div>
<div id="normalization" class="section level2">
<h2 class="hasAnchor">
<a href="#normalization" class="anchor"></a>Normalization</h2>
<p>After cell filtering, the next step is normalization. Normalization is useful for removing cell-specific biases.</p>
<p>Here, we perform the default normalization method in Seurat, which divides gene counts by the total counts for each cell, multiplies this by a scale factor of 10,000, and then log-transforms the result.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">N1469</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">N1469</span><span class="op">)</span></code></pre></div>
</div>
<div id="highly-variable-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#highly-variable-genes" class="anchor"></a>Highly variable genes</h2>
<p>Single cell RNA-seq data is often used for exploring heterogeneity within cell population. To reduce the computational complexity of downstream calculations and also to focus on the true biological signal, a subset of highly variable genes (HVGs) is often selected prior to downstream analyses. One of the most commonly used strategies is to take the top genes with the highest variances across all the cells. The choice of the number of HVGs is fairly arbitrary, with any value from 500 to 5000 considered reasonable.</p>
<p>For this data, we select top 1500 HVGs to be used in downstream analyses such as PCA and UMAP visualization.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">N1469</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html">FindVariableFeatures</a></span><span class="op">(</span><span class="va">N1469</span>, selection.method<span class="op">=</span><span class="st">"vst"</span>, nfeatures<span class="op">=</span><span class="fl">1500</span><span class="op">)</span></code></pre></div>
<p>A mean-variance plot can be produced for visualizing the top variable genes.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">top50</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/VariableFeatures.html">VariableFeatures</a></span><span class="op">(</span><span class="va">N1469</span><span class="op">)</span>, <span class="fl">50</span><span class="op">)</span>
<span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VariableFeaturePlot.html">VariableFeaturePlot</a></span><span class="op">(</span><span class="va">N1469</span><span class="op">)</span>
<span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/LabelPoints.html">LabelPoints</a></span><span class="op">(</span>plot<span class="op">=</span><span class="va">plot1</span>, points<span class="op">=</span><span class="va">top50</span>, repel<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="va">plot2</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="SingleCellWorkshop_files/figure-html/HVG_plot-1.png" alt="A mean-variance plot where top 1500 HVGs are highlighted in red and top 50 HVGs are labelled." width="672"><p class="caption">
A mean-variance plot where top 1500 HVGs are highlighted in red and top 50 HVGs are labelled.
</p>
</div>
<p>Before proceeding to dimensional reduction, we apply a linear transformation to “scale” the data. This data scaling is performed by the <code><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData()</a></code> function, which standardizes the expression of each gene to have a mean expression of 0 and a variance of 1 across all the cells. This step gives equal weight to the genes used in downstream analyses so that highly-expressed genes do not dominate.</p>
<p>By default, the scaling process is only applied to the previously identified 1500 HVGs as these HVGs are used for the downstream analyses.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">N1469</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData</a></span><span class="op">(</span><span class="va">N1469</span><span class="op">)</span></code></pre></div>
</div>
<div id="dimensional-reduction" class="section level2">
<h2 class="hasAnchor">
<a href="#dimensional-reduction" class="anchor"></a>Dimensional reduction</h2>
<p>Dimensional reduction is an essential step in single cell analysis. It summarizes the variances of thousands of genes in a much lower numbers of dimensions, hence reduces computational work in downstream analyses. A simple, highly effective and widely used approach for linear dimensional reduction is principal components analysis (PCA). The top PCs would capture the dominant factors of heterogeneity in the data set.</p>
<p>Here, we perform PCA on the scaled data. By default, only the previously determined 1500 HVGs are used and the first 50 PCs are computed and returned. The PCA results can be visualized in a PCA plot.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">N1469</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">N1469</span>, features<span class="op">=</span><span class="fu"><a href="https://satijalab.org/seurat/reference/VariableFeatures.html">VariableFeatures</a></span><span class="op">(</span><span class="va">N1469</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">N1469</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="SingleCellWorkshop_files/figure-html/PCA-1.png" alt="PCA plot showing the first two principal components of the data." width="672"><p class="caption">
PCA plot showing the first two principal components of the data.
</p>
</div>
<p>Although PCA greatly reduces the dimension of the data from thousands of genes to 50 PCs, it is still difficult to visualize and interpret the 50 PCs at the same time. Therefore, further dimensionality reduction strategies are required to compress the data into 2-3 dimensions for a more intuitive understanding of the data. The two popular non-linear dimensional reduction techniques are <em>t</em>-stochastic neighbor embedding (tSNE) <span class="citation">(Van der Maaten and Hinton 2008)</span> and uniform manifold approximation and projection (UMAP) <span class="citation">(McInnes, Healy, and Melville 2018)</span>.</p>
<p>It is debatable whether the UMAP or tSNE visualization is better. UMAP tends to have more compact visual clusters but reduces resolution within each cluster. The main reason that UMAP has an increasing popularity is that UMAP is much faster than tSNE. Note that both UMAP and tSNE involve a series of randomization steps so setting the seed is critical.</p>
<p>Here we perform both UMAP and tSNE for dimensional reduction and visualization. The top 30 PCs are used as input and a random seed is used for reproducibility.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dimUsed</span> <span class="op">&lt;-</span> <span class="fl">30</span>
<span class="va">N1469</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">N1469</span>, dims<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="va">dimUsed</span>, seed.use<span class="op">=</span><span class="fl">2021</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="va">N1469</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunTSNE.html">RunTSNE</a></span><span class="op">(</span><span class="va">N1469</span>, dims<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="va">dimUsed</span>, seed.use<span class="op">=</span><span class="fl">2021</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">N1469</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span>
<span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">N1469</span>, reduction <span class="op">=</span> <span class="st">"tsne"</span><span class="op">)</span>
<span class="va">plot1</span> <span class="op">+</span> <span class="va">plot2</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="SingleCellWorkshop_files/figure-html/DR_plot-1.png" alt="UMAP and t-SNE visualization" width="864"><p class="caption">
UMAP and t-SNE visualization
</p>
</div>
</div>
<div id="cell-clustering" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-clustering" class="anchor"></a>Cell clustering</h2>
<p>Cell clustering is a procedure in scRNA-seq data analysis to group cells with similar expression profiles. It is an important step for summarizing information and providing biological interpretation of the data. Seurat offers a graph-based clustering approach, which is one of the most popular clustering algorithms due to its flexibilty and scalability for large scRNA-seq datasets.</p>
<p>One of the most commonly asked questions in cell clustering is “how many cell clusters are there in the data?” This question is often hard to answer since we can define as many clusters as we want. In fact, the number of clusters would depend on the biological questions of interest (eg. whether resolution of the major cell types will be sufficient or resolution of subtypes is required). In practice, we often experiment with different resolution in data exploration to obtain the “optimal” resolution that provides the best answer to the questions of our interest.</p>
<p>In Seurat, the cell clustering procedure starts by constructing a KNN graph using the <code><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors()</a></code> function. Here, we use the top 30 PCs as input. Then the Seurat <code><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters()</a></code> function applies the Louvain algorithm (by default) to group cells together. For this particular data, we set the resolution parameter to 0.1. The final clusters can be found using the  function.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">N1469</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">N1469</span>, dims<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="va">dimUsed</span><span class="op">)</span>
<span class="va">N1469</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">N1469</span>, resolution<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></code></pre></div>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 2868
## Number of edges: 109080
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.9600
## Number of communities: 5
## Elapsed time: 0 seconds</code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">N1469</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##    0    1    2    3    4 
## 1699  507  421  213   28</code></pre>
<p>We can visualize the cell clusters in a UMAP plot.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">N1469</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="SingleCellWorkshop_files/figure-html/UMAP-1.png" alt="UMAP visualization where cells are coloured by cell cluster." width="720"><p class="caption">
UMAP visualization where cells are coloured by cell cluster.
</p>
</div>
</div>
<div id="mark-genes-identification" class="section level2">
<h2 class="hasAnchor">
<a href="#mark-genes-identification" class="anchor"></a>Mark genes identification</h2>
<p>The next step after cell clustering is to identify marker genes that drive separation between the cell clusters. Marker genes are usually obtained by performing differential expression analyses between different clusters. In Seurat, the differential expression analysis is performed by the <code><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers()</a></code> function. By default, it identifies positive and negative markers of a single cluster (specified in <code>ident.1</code>), compared to all other cells.</p>
<p>To increase the computational speed, the Seurat <code><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers()</a></code> function only performs DE tests on a subset of genes that satisfy certain thresholds. The <code>min.pct</code> argument requires a gene to be detected at a minimum percentage in either of the two groups of cells, and by default it is set to 10%. The <code>logfc.threshold</code> limits testing to genes of which the log-fold change between the two groups is above a certain level (0.25 by default). By default the <code><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers()</a></code> function performs Wilcoxon Rank Sum tests, but other statistical tests (eg. likelihood-ratio test, <em>t</em>-test) are also available.</p>
<p>Here we find all markers of cluster 1 as follows.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cluster1.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">N1469</span>, ident.1 <span class="op">=</span> <span class="fl">1</span>, min.pct <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cluster1.markers</span><span class="op">)</span></code></pre></div>
<pre><code>##             p_val avg_log2FC pct.1 pct.2 p_val_adj
## CXCL13   0.00e+00       4.17 0.615 0.016  0.00e+00
## DNAJC12  0.00e+00       2.59 0.783 0.027  0.00e+00
## PTHLH    0.00e+00       2.87 0.698 0.016  0.00e+00
## BATF     0.00e+00       2.15 0.684 0.034  0.00e+00
## TFF3     0.00e+00       3.46 0.803 0.010  0.00e+00
## AGR2    3.05e-304       2.22 0.568 0.007 4.85e-300</code></pre>
<p>To find all markers distinguishing cluster 2 from clusters 0 and 1, we use the following lines.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cluster2.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">N1469</span>, ident.1 <span class="op">=</span> <span class="fl">2</span>, ident.2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, min.pct <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cluster2.markers</span><span class="op">)</span></code></pre></div>
<pre><code>##              p_val avg_log2FC pct.1 pct.2 p_val_adj
## ACTG2     0.00e+00       5.09 0.789 0.035   0.0e+00
## CAV1      0.00e+00       3.55 0.891 0.092   0.0e+00
## TPM2      0.00e+00       3.63 0.900 0.068   0.0e+00
## KRT5      0.00e+00       3.79 0.929 0.138   0.0e+00
## KRT14     0.00e+00       5.40 0.969 0.057   0.0e+00
## SERPINB5 2.57e-292       2.26 0.686 0.032  4.1e-288</code></pre>
<p>The Seurat <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers()</a></code> function automates the marker detection process for all clusters. Here we find markers for every cluster compared to all remaining cells, report only the positive ones.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">N1469.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers</a></span><span class="op">(</span><span class="va">N1469</span>, only.pos <span class="op">=</span> <span class="cn">TRUE</span>, min.pct <span class="op">=</span> <span class="fl">0.25</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></code></pre></div>
<p>We select the top 3 marker genes for each of cluster and list them below.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">topMarkers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">N1469.markers</span>, <span class="va">N1469.markers</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span>
<span class="va">top3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">topMarkers</span>, <span class="va">head</span>, n<span class="op">=</span><span class="fl">3</span><span class="op">)</span>
<span class="va">top3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="st">"rbind"</span>, <span class="va">top3</span><span class="op">)</span>
<span class="va">top3</span></code></pre></div>
<pre><code>##            p_val avg_log2FC pct.1 pct.2 p_val_adj cluster     gene
## 0.MMP7         0       4.06 0.957 0.216         0       0     MMP7
## 0.PDZK1IP1     0       2.50 0.950 0.314         0       0 PDZK1IP1
## 0.RPL12        0       1.29 0.999 0.974         0       0    RPL12
## 1.CXCL13       0       4.17 0.615 0.016         0       1   CXCL13
## 1.TFF3         0       3.46 0.803 0.010         0       1     TFF3
## 1.PTHLH        0       2.87 0.698 0.016         0       1    PTHLH
## 2.KRT14        0       5.41 0.969 0.056         0       2    KRT14
## 2.ACTG2        0       5.09 0.789 0.037         0       2    ACTG2
## 2.KRT5         0       3.80 0.929 0.134         0       2     KRT5
## 3.DCN          0       5.76 0.920 0.015         0       3      DCN
## 3.AKR1C1       0       4.97 0.859 0.024         0       3   AKR1C1
## 3.APOD         0       4.56 0.770 0.026         0       3     APOD
## 4.UBE2C        0       2.71 1.000 0.006         0       4    UBE2C
## 4.CDK1         0       1.48 0.679 0.001         0       4     CDK1
## 4.MKI67        0       1.37 0.714 0.000         0       4    MKI67</code></pre>
<p>The expression level of the top marker genes can be overlaid on the UMAP plots for visualization.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">N1469</span>, features <span class="op">=</span> <span class="va">top3</span><span class="op">$</span><span class="va">gene</span>, ncol<span class="op">=</span><span class="fl">3</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="SingleCellWorkshop_files/figure-html/FeaturePlotMarkers-1.png" alt="Top marker expression visualizations on a UMAP plot." width="1056"><p class="caption">
Top marker expression visualizations on a UMAP plot.
</p>
</div>
<p>A dot plot can also be produced for visualizing the top marker genes.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html">DotPlot</a></span><span class="op">(</span><span class="va">N1469</span>, features <span class="op">=</span> <span class="va">top3</span><span class="op">$</span><span class="va">gene</span>, dot.scale <span class="op">=</span> <span class="fl">8</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">RotatedAxis</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="SingleCellWorkshop_files/figure-html/DotPlotMarkers-1.png" alt="A dot plot of top marker genes of each cluster." width="768"><p class="caption">
A dot plot of top marker genes of each cluster.
</p>
</div>
</div>
<div id="cell-type-annotation" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-type-annotation" class="anchor"></a>Cell type annotation</h2>
<p>Interpreting cell clusters in biological context is one of the most challenging tasks in scRNA-seq data analysis. Prior biological knowledge is often required to do so. Marker genes from the literatures or the curated gene sets (e.g., Gene Ontology, KEGG pathways) are the common sources of prior information. Alternatively, we can use published reference datasets where samples are well annotated with cell type information.</p>
<p>Here we use a reference bulk RNA-seq dataset of human mammary gland epithelium from the same study <span class="citation">(Pal et al. 2021)</span>. This reference RNA-seq (GSE161892) data consists of a total of 34 samples from basal, luminal progenitor (LP), mature luminal (ML), and stromal cell types. Differential expression analysis of this reference data was performed using limma-voom and TREAT <span class="citation">(Law et al. 2014; McCarthy and Smyth 2009)</span>. Genes were considered cell type-specific if they were upregulated in one cell type vs all other types. This yielded 515, 323, 765, and 1094 signature genes for basal, LP, ML, and stroma, respectively.</p>
<p>Here we download those signature genes and load them into R.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">url.Signatures</span> <span class="op">&lt;-</span> <span class="st">"https://github.com/SmythLab/scRNAseq-Workshop/raw/main/Data/Human-PosSigGenes.RData"</span>
<span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="va">url.Signatures</span>, destfile<span class="op">=</span><span class="st">"Data/Human-PosSigGenes.RData"</span>, mode<span class="op">=</span><span class="st">"wb"</span><span class="op">)</span> </code></pre></div>
<p>We restrict the signature genes to those expressed in the single cell data.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"Data/Human-PosSigGenes.RData"</span><span class="op">)</span>
<span class="va">HumanSig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Basal<span class="op">=</span><span class="va">Basal</span>, LP<span class="op">=</span><span class="va">LP</span>, ML<span class="op">=</span><span class="va">ML</span>, Str<span class="op">=</span><span class="va">Str</span><span class="op">)</span>
<span class="va">HumanSig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">HumanSig</span>, <span class="va">intersect</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">N1469</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>To associate each cell in the single cell data with the four cell populations in the reference bulk data, we compute the signature scores of the four cell populations for each cell. Here the signature score of a particular cell type is defined as the average expression level of the cell type-specific genes in a given cell.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SigScore</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">HumanSig</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="va">SigScore</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span><span class="op">(</span><span class="va">N1469</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">data</span><span class="op">[</span><span class="va">HumanSig</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="op">]</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>We can visualize the signature scores in both UMAP plots and violin plots.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SigScores</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="st">"cbind"</span>, <span class="va">SigScore</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">SigScores</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Basal"</span>, <span class="st">"LP"</span>, <span class="st">"ML"</span>, <span class="st">"Stroma"</span><span class="op">)</span>
<span class="va">N1469</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">N1469</span><span class="op">@</span><span class="va">meta.data</span>, <span class="va">SigScores</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">N1469</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">SigScores</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="SingleCellWorkshop_files/figure-html/SigScorePlot-1.png" alt="Signature scores." width="960"><p class="caption">
Signature scores.
</p>
</div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">N1469</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">SigScores</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fl">2</span>, pt.size<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="SingleCellWorkshop_files/figure-html/SigScoreViolinPlot-1.png" alt="Violin plots of signature scores." width="1056"><p class="caption">
Violin plots of signature scores.
</p>
</div>
<p>It can be seen that cells in cluster 2 have high expression levels of basal signature genes, suggesting these cells are likely to be basal cells. Likewise, cluster 0 and 4 are LP, cluster 1 is ML and cluster 3 is stroma.</p>
<p>Another way to visualize the correlation between the single cell and the reference bulk dataset is to make a ternary plot, which is useful for studies concerning three major cell populations. For this data, we are particularly interested in assigning each cell to one of the three major epithelial cell populations (basal, LP and ML).</p>
<p>We produce a ternary plot to see which of the three populations the cells are closer to. To measure the similarity between each cell and three populations, we count the numbers of expressed signatures (with at least 1 count) in each cell. The position of each cell on the ternary plot is determined by the numbers of expressed gene signatures of the three populations in that cell.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">TN</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fl">0L</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">N1469</span><span class="op">)</span>, <span class="fl">3L</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">TN</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"LP"</span>, <span class="st">"ML"</span>, <span class="st">"Basal"</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">TN</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="va">TN</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">N1469</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span><span class="op">[</span><span class="va">HumanSig</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0L</span><span class="op">)</span>
<span class="op">}</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">vcd</span><span class="op">)</span>
<span class="va">col.p</span> <span class="op">&lt;-</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/hue_pal.html">hue_pal</a></span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nlevels.html">nlevels</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">N1469</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/vcd/man/ternaryplot.html">ternaryplot</a></span><span class="op">(</span><span class="va">TN</span>, cex<span class="op">=</span><span class="fl">0.2</span>, pch<span class="op">=</span><span class="fl">16</span>, col<span class="op">=</span><span class="va">col.p</span><span class="op">[</span><span class="fu"><a href="https://satijalab.org/seurat/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">N1469</span><span class="op">)</span><span class="op">]</span>, grid<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="SingleCellWorkshop_files/figure-html/TernaryPlot-1.png" alt="Violin plots of signature scores." width="528"><p class="caption">
Violin plots of signature scores.
</p>
</div>
</div>
</div>
<div id="integration-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#integration-analysis" class="anchor"></a>Integration analysis</h1>
<div id="batch-effects" class="section level2">
<h2 class="hasAnchor">
<a href="#batch-effects" class="anchor"></a>Batch effects</h2>
<p>Comprehensive single cell RNA-seq experiments usually consist of multiple samples, which may be collected and prepared at different times, by different operators or with different reagents. These inconsistencies are referred to as “batch effects” and thye often lead to systematic differences in the expression profiles of the scRNA-seq data. Batch effects can be problematic as they distort the genuine biological differences and may mislead the interpretation of the data. Therefore, integration methods that adjust for “batch effects” are highly essential in multi-sample single cell RNA-seq analysis.</p>
<p>Here, we will demonstrate how to perform an integration analysis in Seurat.</p>
</div>
<div id="read-in-the-data-1" class="section level2">
<h2 class="hasAnchor">
<a href="#read-in-the-data-1" class="anchor"></a>Read in the data</h2>
<p>We use 10X single cell RNA-seq profiles of normal epithelial cells from five individual human. We first read in the raw data of all five samples for the integration analysis.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"N1469"</span>, <span class="st">"N0280"</span>, <span class="st">"N1105"</span>, <span class="st">"N0230"</span>, <span class="st">"N0123"</span><span class="op">)</span>
<span class="va">RawData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">RawData</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">N1469.data</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">2</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">Samples</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">RawData</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X.html">Read10X</a></span><span class="op">(</span>data.dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Data/"</span>, <span class="va">Samples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">RawData</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">Samples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">RawData</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, sep<span class="op">=</span><span class="st">"_"</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">RawData</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">Samples</span></code></pre></div>
<p>Then we create a list of 5 Seurat objects (one for each human sample). The same cell filtering criteria is applied to all the samples.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SeuratList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">Samples</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="va">SeuratList</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span>counts<span class="op">=</span><span class="va">RawData</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, project<span class="op">=</span><span class="va">Samples</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, min.cells<span class="op">=</span><span class="fl">3</span>, min.features<span class="op">=</span><span class="fl">200</span><span class="op">)</span>
    <span class="va">SeuratList</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="st">"percent.mt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">SeuratList</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span>
    <span class="va">SeuratList</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">SeuratList</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, subset <span class="op">=</span> <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">500</span> <span class="op">&amp;</span> <span class="va">percent.mt</span> <span class="op">&lt;</span> <span class="fl">20</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<p>We perform all the pre-processing steps (normalization, defining HVGs and scaling the data) for each individual sample.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SeuratList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">SeuratList</span>, <span class="va">NormalizeData</span><span class="op">)</span>
<span class="va">SeuratList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">SeuratList</span>, <span class="va">FindVariableFeatures</span>, selection.method<span class="op">=</span><span class="st">"vst"</span>, nfeatures<span class="op">=</span><span class="fl">1500</span><span class="op">)</span>
<span class="va">SeuratList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">SeuratList</span>, <span class="va">ScaleData</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">SeuratList</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">Samples</span></code></pre></div>
</div>
<div id="data-integration" class="section level2">
<h2 class="hasAnchor">
<a href="#data-integration" class="anchor"></a>Data integration</h2>
<p>Seurat implements an anchor-based method for integrating multiple single cell RNA-seq samples. The method first identifies pairs of cells between different samples (“anchors”) that are in a matched biological state. These “anchors” are then used to harmonize datasets into a single reference, of which the labels and data can be projected onto query datasets <span class="citation">(Stuart et al. 2019)</span>.</p>
<p>We identify anchors using the <code><a href="https://satijalab.org/seurat/reference/FindIntegrationAnchors.html">FindIntegrationAnchors()</a></code> function, which takes a list of Seurat objects as input.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Anchors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindIntegrationAnchors.html">FindIntegrationAnchors</a></span><span class="op">(</span><span class="va">SeuratList</span><span class="op">)</span></code></pre></div>
<p>We then use these anchors to integrate the five datasets together with <code><a href="https://satijalab.org/seurat/reference/IntegrateData.html">IntegrateData()</a></code>. This step creates an  data assay.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">NormEpi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/IntegrateData.html">IntegrateData</a></span><span class="op">(</span><span class="va">Anchors</span><span class="op">)</span></code></pre></div>
<p>Now we can run a single integrated analysis on all cells. The downstream analysis will be performed on corrected data (the <code>integrated</code> data assay). The original unmodified data still resides in the <code>RNA</code> assay.</p>
</div>
<div id="dimensional-reduction-1" class="section level2">
<h2 class="hasAnchor">
<a href="#dimensional-reduction-1" class="anchor"></a>Dimensional reduction</h2>
<p>The dimensional reduction step for the integrated data is the same as for a single sample. We first scale the data.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">NormEpi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData</a></span><span class="op">(</span><span class="va">NormEpi</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p>Then we perform dimensional reduction (PCA and UMAP) for the integrated data.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dimUsed</span> <span class="op">&lt;-</span> <span class="fl">30</span>
<span class="va">NormEpi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span><span class="va">NormEpi</span>, npcs<span class="op">=</span><span class="va">dimUsed</span>, verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="va">NormEpi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span><span class="va">NormEpi</span>, dims<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="va">dimUsed</span>, seed.use<span class="op">=</span><span class="fl">2021</span><span class="op">)</span></code></pre></div>
<p>The UMAP visualization of the integrated data is shown below.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">NormEpi</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by<span class="op">=</span><span class="st">"orig.ident"</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="SingleCellWorkshop_files/figure-html/UMAP_Sample-1.png" alt="UMAP visualization of the integrated data where cells are coloured by patient." width="768"><p class="caption">
UMAP visualization of the integrated data where cells are coloured by patient.
</p>
</div>
<p>From the UMAP it can be seen that cells are not clustered by sample, indicating the batch effects between different samples have been removed. It is also noticable that each of the big clusters contains a mixture of cells from different samples, suggesting these five samples share a similar cell type composition.</p>
</div>
<div id="cell-clustering-1" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-clustering-1" class="anchor"></a>Cell clustering</h2>
<p>We then proceed to cell clustering of the integrated data, which is done in the same way as before.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">NormEpi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span><span class="va">NormEpi</span>, dims<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="va">dimUsed</span><span class="op">)</span>
<span class="va">NormEpi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span><span class="va">NormEpi</span>, resolution<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></code></pre></div>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 13644
## Number of edges: 572889
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.9642
## Number of communities: 6
## Elapsed time: 2 seconds</code></pre>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">NormEpi</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##    0    1    2    3    4    5 
## 5520 5446 2040  418  153   67</code></pre>
<p>The clustering results are shown in the UMAP plot below.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">NormEpi</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, label<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="SingleCellWorkshop_files/figure-html/UMAP_Cluster-1.png" alt="UMAP visualization of the integrated data where cells are coloured by cluster." width="768"><p class="caption">
UMAP visualization of the integrated data where cells are coloured by cluster.
</p>
</div>
</div>
<div id="cell-type-annotation-1" class="section level2">
<h2 class="hasAnchor">
<a href="#cell-type-annotation-1" class="anchor"></a>Cell type annotation</h2>
<p>The same human reference data is used for cell type annotation. The signature scores of the four cell populations are calculated for each cell in the integrated data.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">HumanSig2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Basal<span class="op">=</span><span class="va">Basal</span>, LP<span class="op">=</span><span class="va">LP</span>, ML<span class="op">=</span><span class="va">ML</span>, Str<span class="op">=</span><span class="va">Str</span><span class="op">)</span>
<span class="va">HumanSig2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">HumanSig2</span>, <span class="va">intersect</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">NormEpi</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SigScore2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">HumanSig2</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="va">SigScore2</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span><span class="op">(</span><span class="va">NormEpi</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">data</span><span class="op">[</span><span class="va">HumanSig</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="op">]</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SigScores2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="st">"cbind"</span>, <span class="va">SigScore2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">SigScores2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Basal"</span>, <span class="st">"LP"</span>, <span class="st">"ML"</span>, <span class="st">"Stroma"</span><span class="op">)</span>
<span class="va">NormEpi</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">NormEpi</span><span class="op">@</span><span class="va">meta.data</span>, <span class="va">SigScores2</span><span class="op">)</span></code></pre></div>
<p>We then produce feature plots and violin plots to visualize the signature scores.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">NormEpi</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">SigScores2</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="SingleCellWorkshop_files/figure-html/SigScorePlot2-1.png" alt="Signature scores for the integrated data." width="960"><p class="caption">
Signature scores for the integrated data.
</p>
</div>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">NormEpi</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">SigScores2</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fl">2</span>, pt.size<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="SingleCellWorkshop_files/figure-html/SigScoreViolinPlot2-1.png" alt="Violin plots of signature scores." width="1056"><p class="caption">
Violin plots of signature scores.
</p>
</div>
<p>It can be seen that cells in cluster 2 have high expression levels of basal signature genes, suggesting these cells are likely to be basal cells. Likewise, cluster 0 and 5 are LP or LP-like, cluster 1 and 4 are ML or ML-like, and cluster 3 is stroma.</p>
<p>A ternary plot can also be generated in the same way to examine the correlation between the cells and the three major epithelial populations.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">TN2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fl">0L</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">NormEpi</span><span class="op">)</span>, <span class="fl">3L</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">TN2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"LP"</span>, <span class="st">"ML"</span>, <span class="st">"Basal"</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">TN2</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
    <span class="va">TN2</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">NormEpi</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span><span class="op">[</span><span class="va">HumanSig2</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0L</span><span class="op">)</span>
<span class="op">}</span>
<span class="va">col.p</span> <span class="op">&lt;-</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/hue_pal.html">hue_pal</a></span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nlevels.html">nlevels</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">NormEpi</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/vcd/man/ternaryplot.html">ternaryplot</a></span><span class="op">(</span><span class="va">TN2</span>, cex<span class="op">=</span><span class="fl">0.2</span>, pch<span class="op">=</span><span class="fl">16</span>, col<span class="op">=</span><span class="va">col.p</span><span class="op">[</span><span class="fu"><a href="https://satijalab.org/seurat/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">NormEpi</span><span class="op">)</span><span class="op">]</span>, grid<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="SingleCellWorkshop_files/figure-html/TernaryPlot2-1.png" alt="Violin plots of signature scores for the integrated data." width="576"><p class="caption">
Violin plots of signature scores for the integrated data.
</p>
</div>
</div>
</div>
<div id="differential-expression-analysis-by-pseudo-bulk" class="section level1">
<h1 class="hasAnchor">
<a href="#differential-expression-analysis-by-pseudo-bulk" class="anchor"></a>Differential expression analysis by pseudo-bulk</h1>
<div id="biological-variation" class="section level2">
<h2 class="hasAnchor">
<a href="#biological-variation" class="anchor"></a>Biological variation</h2>
<p>In RNA-seq experiments with replicate samples, biological variation represents the variation of the underlying gene abundances between the replicates. Accounting for biological variation is crucial in RNA-seq DE analysis. Many statistically rigorous methods and software packages have been developed for estimating biological variation and incorporating it into the DE analysis (e.g., edgeR, limma, DESeq2 etc.).</p>
<p>For single cell RNA-seq experiments with biological replicates, one should also take into account the biological variation between different replicate samples where cells are obtained. However, applying bulk RNA-seq DE methods directly to single cell RNA-seq data is not appropriate as each cell is not independent biological replicate. One of the commonly used methods to account for biological variation between replicates is the “pseudo-bulk” approach, where read counts from all cells with the same combination of label (e.g., cluster) and sample are summed together <span class="citation">(Tung et al. 2017)</span>.</p>
</div>
<div id="creating-pseudo-bulk-samples" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-pseudo-bulk-samples" class="anchor"></a>Creating pseudo-bulk samples</h2>
<p>We first extract the cluster and sample information from the integrated Seurat object and combine them together. The numbers of cells under different combinations of cell cluster and sample are shown below.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sampID</span> <span class="op">&lt;-</span> <span class="va">NormEpi</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">orig.ident</span>
<span class="va">clustID</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">NormEpi</span><span class="op">)</span>
<span class="va">sampClust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">sampID</span>, <span class="va">clustID</span>, sep<span class="op">=</span><span class="st">"_Clst"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">sampClust</span><span class="op">)</span></code></pre></div>
<pre><code>## sampClust
## N0123_Clst0 N0123_Clst1 N0123_Clst2 N0123_Clst3 N0123_Clst4 N0123_Clst5 N0230_Clst0 
##         839        1563         366          12          41          19        1033 
## N0230_Clst1 N0230_Clst2 N0230_Clst3 N0230_Clst4 N0230_Clst5 N0280_Clst0 N0280_Clst1 
##         561         156          77          15          15         346         392 
## N0280_Clst2 N0280_Clst3 N0280_Clst4 N0280_Clst5 N1105_Clst0 N1105_Clst1 N1105_Clst2 
##         316          39          30           3        1597        2469         784 
## N1105_Clst3 N1105_Clst4 N1105_Clst5 N1469_Clst0 N1469_Clst1 N1469_Clst2 N1469_Clst3 
##          80          22           1        1705         461         418         210 
## N1469_Clst4 N1469_Clst5 
##          45          29</code></pre>
<p>We then form pseudo-bulk samples by aggregating the read counts for each gene by cluster-sample combination.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rawCounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">NormEpi</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">RNA</span><span class="op">@</span><span class="va">counts</span><span class="op">)</span>
<span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rowsum.html">rowsum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">rawCounts</span><span class="op">)</span>, group<span class="op">=</span><span class="va">sampClust</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>We create a <code>DGEList</code> object using the pseudo-bulk samples and proceed to the edgeR DE analysis pipeline <span class="citation">(Chen, Lun, and Smyth 2016)</span>.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span><span class="op">)</span>
<span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html">DGEList</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>
<span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"^.*_Clst"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">Patient</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"_.*$"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span>, level<span class="op">=</span><span class="va">Samples</span><span class="op">)</span></code></pre></div>
</div>
<div id="filtering" class="section level2">
<h2 class="hasAnchor">
<a href="#filtering" class="anchor"></a>Filtering</h2>
<p>The library size (total number of reads) of each pseudo-bulk sample highly depends on the number of cells in that sample. We remove the samples with library size less than 50,000.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">keep.sample</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">lib.size</span> <span class="op">&gt;</span> <span class="fl">5e4</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">keep.sample</span><span class="op">)</span></code></pre></div>
<pre><code>## keep.sample
## FALSE  TRUE 
##     2    28</code></pre>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span>,<span class="va">keep.sample</span><span class="op">]</span></code></pre></div>
<p>The gene filtering is performed by the <code>filterByExpr</code> function in edgeR.</p>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">keep.gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html">filterByExpr</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">keep.gene</span><span class="op">)</span></code></pre></div>
<pre><code>## keep.gene
## FALSE  TRUE 
##  8775 10454</code></pre>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="va">keep.gene</span>,,keep<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span></code></pre></div>
</div>
<div id="normalization-1" class="section level2">
<h2 class="hasAnchor">
<a href="#normalization-1" class="anchor"></a>Normalization</h2>
<p>We perform TMM normalization <span class="citation">(Robinson and Oshlack 2010)</span> to adjust for the compositional biases between the samples. The normalization factors are listed below.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span>
<span class="va">y</span><span class="op">$</span><span class="va">samples</span></code></pre></div>
<pre><code>##             group lib.size norm.factors Patient
## N0123_Clst0     0  4252753        0.914   N0123
## N0123_Clst1     1  5913960        1.033   N0123
## N0123_Clst2     2   933292        0.870   N0123
## N0123_Clst4     4   144308        1.135   N0123
## N0123_Clst5     5   155185        1.155   N0123
## N0230_Clst0     0  4261533        1.110   N0230
## N0230_Clst1     1  1935170        1.140   N0230
## N0230_Clst2     2   452640        1.031   N0230
## N0230_Clst3     3   214246        1.115   N0230
## N0230_Clst4     4    67372        1.410   N0230
## N0230_Clst5     5    80893        1.437   N0230
## N0280_Clst0     0  3774944        0.729   N0280
## N0280_Clst1     1  2031958        1.014   N0280
## N0280_Clst2     2  1382642        0.875   N0280
## N0280_Clst3     3   188620        1.039   N0280
## N0280_Clst4     4   180587        1.185   N0280
## N0280_Clst5     5    55465        1.251   N0280
## N1105_Clst0     0 18342441        0.767   N1105
## N1105_Clst1     1 15781150        0.968   N1105
## N1105_Clst2     2  3730398        0.778   N1105
## N1105_Clst3     3   430622        0.908   N1105
## N1105_Clst4     4   172690        1.219   N1105
## N1469_Clst0     0 10459180        0.859   N1469
## N1469_Clst1     1  2315185        0.923   N1469
## N1469_Clst2     2  1745234        0.831   N1469
## N1469_Clst3     3  1170267        0.776   N1469
## N1469_Clst4     4   234042        0.933   N1469
## N1469_Clst5     5   261094        1.034   N1469</code></pre>
</div>
<div id="data-exploration" class="section level2">
<h2 class="hasAnchor">
<a href="#data-exploration" class="anchor"></a>Data exploration</h2>
<p>We produce an MDS plot to visualize the relationship between all the pseudo-bulk samples.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">group</span><span class="op">)</span>
<span class="va">Patient</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">Patient</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="fu">plotMDS</span><span class="op">(</span><span class="va">y</span>, pch<span class="op">=</span><span class="fl">16</span>, col<span class="op">=</span><span class="va">col.p</span><span class="op">[</span><span class="va">Cluster</span><span class="op">]</span>, main<span class="op">=</span><span class="st">"By cluster"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="st">"bottomleft"</span>, legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Cluster"</span>, <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">Cluster</span><span class="op">)</span><span class="op">)</span>, pch<span class="op">=</span><span class="fl">16</span>, col<span class="op">=</span><span class="va">col.p</span><span class="op">)</span>
<span class="fu">plotMDS</span><span class="op">(</span><span class="va">y</span>, pch<span class="op">=</span><span class="fl">16</span>, col<span class="op">=</span><span class="va">Patient</span>, main<span class="op">=</span><span class="st">"By sample"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span><span class="op">(</span><span class="st">"bottomleft"</span>, legend<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">Patient</span><span class="op">)</span>, pch<span class="op">=</span><span class="fl">16</span>, col<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="SingleCellWorkshop_files/figure-html/MDS-1.png" alt="MDS plots of the pseudo-bulk samples." width="1056"><p class="caption">
MDS plots of the pseudo-bulk samples.
</p>
</div>
<p>From the MDS plot it can be seen that the pseudo-bulk samples are grouped by cluster defined at the single cell level. We also notice that samples from cluster 0 and 5 have similar expression pattern, whereas samples from cluster 1 and 4 are close to each other.</p>
<p>We use the cluster and sample information to create a design matrix as follows.</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">Cluster</span> <span class="op">+</span> <span class="va">Patient</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] "Cluster0"     "Cluster1"     "Cluster2"     "Cluster3"     "Cluster4"    
##  [6] "Cluster5"     "PatientN0280" "PatientN1105" "PatientN0230" "PatientN0123"</code></pre>
</div>
<div id="dispersion-estimation" class="section level2">
<h2 class="hasAnchor">
<a href="#dispersion-estimation" class="anchor"></a>Dispersion estimation</h2>
<p>We estimate negative binomial dispersions and quasi-likelihood dispersions as follows.</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html">estimateDisp</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span>
<span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html">glmQLFit</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span></code></pre></div>
<p>The dispersions can be examined in the following plots.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/plotBCV.html">plotBCV</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/plotQLDisp.html">plotQLDisp</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="SingleCellWorkshop_files/figure-html/plotDisp-1.png" alt="BCV and quasi-likelihood dispersion plots." width="1056"><p class="caption">
BCV and quasi-likelihood dispersion plots.
</p>
</div>
</div>
<div id="differential-expression" class="section level2">
<h2 class="hasAnchor">
<a href="#differential-expression" class="anchor"></a>Differential expression</h2>
<p>For the differential expression analysis, we are particularly interested in the genes that are up-regulated in each cluster compared with all the other clusters. Therefore, we create a contrast matrix as follows. Each column of the matrix represents a testing constrast that compares one cluster with the average of all the other clusters.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ncls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nlevels.html">nlevels</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">NormEpi</span><span class="op">)</span><span class="op">)</span>
<span class="va">contr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">ncls</span><span class="op">)</span>, <span class="va">ncls</span>, <span class="va">ncls</span><span class="op">)</span>,
                <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span><span class="op">-</span><span class="va">ncls</span>, <span class="va">ncls</span><span class="op">)</span> <span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="va">contr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">contr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">contr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Cluster"</span>, <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">NormEpi</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">contr</span></code></pre></div>
<pre><code>##              Cluster0 Cluster1 Cluster2 Cluster3 Cluster4 Cluster5
## Cluster0          1.0     -0.2     -0.2     -0.2     -0.2     -0.2
## Cluster1         -0.2      1.0     -0.2     -0.2     -0.2     -0.2
## Cluster2         -0.2     -0.2      1.0     -0.2     -0.2     -0.2
## Cluster3         -0.2     -0.2     -0.2      1.0     -0.2     -0.2
## Cluster4         -0.2     -0.2     -0.2     -0.2      1.0     -0.2
## Cluster5         -0.2     -0.2     -0.2     -0.2     -0.2      1.0
## PatientN0280      0.0      0.0      0.0      0.0      0.0      0.0
## PatientN1105      0.0      0.0      0.0      0.0      0.0      0.0
## PatientN0230      0.0      0.0      0.0      0.0      0.0      0.0
## PatientN0123      0.0      0.0      0.0      0.0      0.0      0.0</code></pre>
<p>We then perform quasi-likelihood F-test under each contrast. In order to focus on genes with strong DE signal, we test for the differences above a fold-change threshold of 1.3. This is done by calling the <code>glmTreat</code> function in edgeR. The results of all the tests are stored as a list of <code>DGELRT</code> objects.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">ncls</span><span class="op">)</span> <span class="va">test</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmTreat.html">glmTreat</a></span><span class="op">(</span><span class="va">fit</span>, contrast<span class="op">=</span><span class="va">contr</span><span class="op">[</span>,<span class="va">i</span><span class="op">]</span>, lfc<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">1.3</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">test</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">contr</span><span class="op">)</span></code></pre></div>
<p>The number of DE genes under each comparison can be viewed by the <code>decideTestsDGE</code> function.</p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dtest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="va">test</span>, <span class="va">decideTestsDGE</span><span class="op">)</span>, <span class="va">summary</span><span class="op">)</span>
<span class="va">dtest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="st">"cbind"</span>, <span class="va">dtest</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">dtest</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">contr</span><span class="op">)</span>
<span class="va">dtest</span></code></pre></div>
<pre><code>##        Cluster0 Cluster1 Cluster2 Cluster3 Cluster4 Cluster5
## Down        574      919     1506     1109     1481      497
## NotSig     8599     7161     6685     7692     7401     9380
## Up         1281     2374     2263     1653     1572      577</code></pre>
<p>We can view the top DE results using the <code>topTags</code> function. For example, to see the top 20 DE genes of cluster 2 vs all the other clusters, we use the following line.</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html">topTags</a></span><span class="op">(</span><span class="va">test</span><span class="op">$</span><span class="va">Cluster2</span>, n<span class="op">=</span><span class="fl">20L</span><span class="op">)</span></code></pre></div>
<pre><code>## Coefficient:  -0.2*Cluster0 -0.2*Cluster1 1*Cluster2 -0.2*Cluster3 -0.2*Cluster4 -0.2*Cluster5 
##          logFC unshrunk.logFC logCPM   PValue      FDR
## ACTG2     7.72           7.73   9.21 2.82e-26 2.95e-22
## SERPINB5  6.28           6.30   6.56 3.58e-25 1.87e-21
## KRT5      6.51           6.52   8.98 2.57e-24 7.20e-21
## TPM2      5.77           5.78   8.27 2.76e-24 7.20e-21
## DLL1      5.38           5.39   5.96 9.51e-24 1.99e-20
## DKK3      5.77           5.78   6.70 1.24e-23 2.17e-20
## CD200     6.11           6.13   6.52 1.57e-23 2.34e-20
## KRT14     7.39           7.39  10.84 1.16e-22 1.51e-19
## TNS4      4.96           4.97   6.25 1.76e-21 2.04e-18
## ACTA2     6.22           6.23   8.45 2.66e-21 2.78e-18
## GPC1      4.65           4.66   6.31 3.87e-21 3.68e-18
## NRP2      5.92           5.96   6.08 4.25e-21 3.70e-18
## COL17A1   8.93          16.15   5.34 4.69e-21 3.77e-18
## CCND2     7.07          10.60   5.68 5.44e-21 4.06e-18
## CAV1      4.76           4.76   8.32 6.77e-21 4.72e-18
## MSRB3     5.27           5.29   5.38 7.63e-21 4.99e-18
## TUBB2B    4.44           4.44   7.58 1.00e-20 6.17e-18
## OXTR      5.35           5.37   6.25 1.14e-20 6.61e-18
## DST       3.84           3.84   8.71 2.04e-20 1.12e-17
## FAM126A   4.41           4.42   5.87 2.87e-20 1.50e-17</code></pre>
<p>We can extract the top 15 marker genes (up-regulated only) of each cluster as follows.</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">top</span> <span class="op">&lt;-</span> <span class="fl">15</span>
<span class="va">pseudoMakers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="va">ncls</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">res</span> <span class="op">&lt;-</span> <span class="va">test</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">table</span>
    <span class="va">ord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">PValue</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/sign.html">sign</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span><span class="op">)</span>
    <span class="va">pseudoMakers</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">[</span><span class="va">ord</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">top</span><span class="op">]</span><span class="op">]</span>
<span class="op">}</span>
<span class="va">Markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">pseudoMakers</span><span class="op">)</span>
<span class="va">Markers</span> <span class="op">&lt;-</span> <span class="va">Markers</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span><span class="op">(</span><span class="va">Markers</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<p>Then we create a heatmap using <code>pheatmap</code> for those top marker genes. The standardized log-CPM values of the marker genes are used for the expression levels on the heatmap. Genes and samples on the heatmap are reordered by the Ward method of hierarchical clustering, which correctly groups samples by cell cluster.</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span>
<span class="va">lcpm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html">cpm</a></span><span class="op">(</span><span class="va">y</span>, log<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
<span class="va">expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">lcpm</span><span class="op">[</span><span class="va">Markers</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>Cluster<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Cluster "</span>, <span class="va">Cluster</span><span class="op">)</span>, Patient<span class="op">=</span><span class="va">Patient</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">annot</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">expr</span><span class="op">)</span>
<span class="va">ann_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Cluster<span class="op">=</span><span class="va">col.p</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">ncls</span><span class="op">]</span>, Patient<span class="op">=</span><span class="va">col.p</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">Samples</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">ann_colors</span><span class="op">$</span><span class="va">Cluster</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Cluster "</span>, <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">NormEpi</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">ann_colors</span><span class="op">$</span><span class="va">Patient</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">Samples</span>
<span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span><span class="op">(</span><span class="va">expr</span>, color<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>,<span class="st">"white"</span>,<span class="st">"red"</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>, border_color<span class="op">=</span><span class="st">"NA"</span>,
    breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2</span>,<span class="fl">2</span>,length.out<span class="op">=</span><span class="fl">101</span><span class="op">)</span>, cluster_cols<span class="op">=</span><span class="cn">TRUE</span>, scale<span class="op">=</span><span class="st">"none"</span>, fontsize_row<span class="op">=</span><span class="fl">7</span>,
    show_colnames<span class="op">=</span><span class="cn">FALSE</span>, treeheight_row<span class="op">=</span><span class="fl">70</span>, treeheight_col<span class="op">=</span><span class="fl">70</span>, cutree_cols<span class="op">=</span><span class="va">ncls</span>,
    clustering_method<span class="op">=</span><span class="st">"ward.D2"</span>, annotation_col<span class="op">=</span><span class="va">annot</span>, annotation_colors<span class="op">=</span><span class="va">ann_colors</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="SingleCellWorkshop_files/figure-html/HeatmapPB-1.png" alt="Heat map of the top marker genes." width="720"><p class="caption">
Heat map of the top marker genes.
</p>
</div>
</div>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.1.0 (2021-05-18)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.2 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
##  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] grid      stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] pheatmap_1.0.12    edgeR_3.35.0       limma_3.49.4       vcd_1.4-8         
## [5] SeuratObject_4.0.2 Seurat_4.0.3       BiocStyle_2.21.3  
## 
## loaded via a namespace (and not attached):
##   [1] Rtsne_0.15            colorspace_2.0-2      deldir_0.2-10        
##   [4] ellipsis_0.3.2        ggridges_0.5.3        rprojroot_2.0.2      
##   [7] fs_1.5.0              spatstat.data_2.1-0   farver_2.1.0         
##  [10] leiden_0.3.9          listenv_0.8.0         ggrepel_0.9.1        
##  [13] RSpectra_0.16-0       fansi_0.5.0           codetools_0.2-18     
##  [16] splines_4.1.0         cachem_1.0.5          knitr_1.33           
##  [19] polyclip_1.10-0       jsonlite_1.7.2        ica_1.0-2            
##  [22] cluster_2.1.2         png_0.1-7             uwot_0.1.10          
##  [25] shiny_1.6.0           sctransform_0.3.2     spatstat.sparse_2.0-0
##  [28] BiocManager_1.30.16   compiler_4.1.0        httr_1.4.2           
##  [31] Matrix_1.3-4          fastmap_1.1.0         lazyeval_0.2.2       
##  [34] later_1.2.0           htmltools_0.5.1.1     tools_4.1.0          
##  [37] igraph_1.2.6          gtable_0.3.0          glue_1.4.2           
##  [40] RANN_2.6.1            reshape2_1.4.4        dplyr_1.0.7          
##  [43] Rcpp_1.0.7            scattermore_0.7       jquerylib_0.1.4      
##  [46] pkgdown_1.6.1         vctrs_0.3.8           nlme_3.1-152         
##  [49] lmtest_0.9-38         xfun_0.25             stringr_1.4.0        
##  [52] globals_0.14.0        mime_0.11             miniUI_0.1.1.1       
##  [55] lifecycle_1.0.0       irlba_2.3.3           goftest_1.2-2        
##  [58] future_1.21.0         MASS_7.3-54           zoo_1.8-9            
##  [61] scales_1.1.1          spatstat.core_2.3-0   ragg_1.1.3           
##  [64] promises_1.2.0.1      spatstat.utils_2.2-0  parallel_4.1.0       
##  [67] RColorBrewer_1.1-2    yaml_2.2.1            memoise_2.0.0        
##  [70] reticulate_1.20       pbapply_1.4-3         gridExtra_2.3        
##  [73] ggplot2_3.3.5         sass_0.4.0            rpart_4.1-15         
##  [76] stringi_1.7.3         highr_0.9             desc_1.3.0           
##  [79] rlang_0.4.11          pkgconfig_2.0.3       systemfonts_1.0.2    
##  [82] matrixStats_0.60.0    evaluate_0.14         lattice_0.20-44      
##  [85] tensor_1.5            ROCR_1.0-11           purrr_0.3.4          
##  [88] labeling_0.4.2        patchwork_1.1.1       htmlwidgets_1.5.3    
##  [91] cowplot_1.1.1         tidyselect_1.1.1      parallelly_1.27.0    
##  [94] RcppAnnoy_0.0.19      plyr_1.8.6            magrittr_2.0.1       
##  [97] bookdown_0.23         R6_2.5.0              generics_0.1.0       
## [100] withr_2.4.2           mgcv_1.8-36           pillar_1.6.2         
## [103] fitdistrplus_1.1-5    abind_1.4-5           survival_3.2-12      
## [106] tibble_3.1.3          future.apply_1.8.1    crayon_1.4.1         
## [109] KernSmooth_2.23-20    utf8_1.2.2            spatstat.geom_2.2-2  
## [112] plotly_4.9.4.1        rmarkdown_2.10        locfit_1.5-9.4       
## [115] data.table_1.14.0     digest_0.6.27         xtable_1.8-4         
## [118] tidyr_1.1.3           httpuv_1.6.1          textshaping_0.3.5    
## [121] munsell_0.5.0         viridisLite_0.4.0     bslib_0.2.5.1</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-chen2016reads" class="csl-entry">
Chen, Yunshun, Aaron TL Lun, and Gordon K Smyth. 2016. <span>“From Reads to Genes to Pathways: Differential Expression Analysis of RNA-Seq Experiments Using Rsubread and the edgeR Quasi-Likelihood Pipeline.”</span> <em>F1000Research</em> 5.
</div>
<div id="ref-ilicic2016classification" class="csl-entry">
Ilicic, Tomislav, Jong Kyoung Kim, Aleksandra A Kolodziejczyk, Frederik Otzen Bagger, Davis James McCarthy, John C Marioni, and Sarah A Teichmann. 2016. <span>“Classification of Low Quality Cells from Single-Cell RNA-Seq Data.”</span> <em>Genome Biology</em> 17 (1): 1–15.
</div>
<div id="ref-law2014voom" class="csl-entry">
Law, Charity W, Yunshun Chen, Wei Shi, and Gordon K Smyth. 2014. <span>“Voom: Precision Weights Unlock Linear Model Analysis Tools for RNA-Seq Read Counts.”</span> <em>Genome Biology</em> 15 (2): 1–17.
</div>
<div id="ref-mccarthy2009testing" class="csl-entry">
McCarthy, Davis J, and Gordon K Smyth. 2009. <span>“Testing Significance Relative to a Fold-Change Threshold Is a TREAT.”</span> <em>Bioinformatics</em> 25 (6): 765–71.
</div>
<div id="ref-mcinnes2018umap" class="csl-entry">
McInnes, Leland, John Healy, and James Melville. 2018. <span>“Umap: Uniform Manifold Approximation and Projection for Dimension Reduction.”</span> <em>arXiv Preprint arXiv:1802.03426</em>.
</div>
<div id="ref-pal2021scrnaseqatlas" class="csl-entry">
Pal, Bhupinder, Yunshun Chen, François Vaillant, Bianca D. Capaldo, Rachel Joyce, Xiaoyu Song, Vanessa Bryant, et al. 2021. <span>“A Single Cell <span>RNA</span> Atlas of Human Breast Spanning Normal, Preneoplastic and Tumorigenic States.”</span> <em>EMBO Journal</em>, Accepted 17 March 2021.
</div>
<div id="ref-robinson2010scaling" class="csl-entry">
Robinson, Mark D, and Alicia Oshlack. 2010. <span>“A Scaling Normalization Method for Differential Expression Analysis of RNA-Seq Data.”</span> <em>Genome Biology</em> 11 (3): 1–9.
</div>
<div id="ref-satija2015spatial" class="csl-entry">
Satija, Rahul, Jeffrey A Farrell, David Gennert, Alexander F Schier, and Aviv Regev. 2015. <span>“Spatial Reconstruction of Single-Cell Gene Expression Data.”</span> <em>Nature Biotechnology</em> 33 (5): 495–502.
</div>
<div id="ref-stuart2019comprehensive" class="csl-entry">
Stuart, Tim, Andrew Butler, Paul Hoffman, Christoph Hafemeister, Efthymia Papalexi, William M Mauck III, Yuhan Hao, Marlon Stoeckius, Peter Smibert, and Rahul Satija. 2019. <span>“Comprehensive Integration of Single-Cell Data.”</span> <em>Cell</em> 177 (7): 1888–1902.
</div>
<div id="ref-tung2017batch" class="csl-entry">
Tung, Po-Yuan, John D Blischak, Chiaowen Joyce Hsiao, David A Knowles, Jonathan E Burnett, Jonathan K Pritchard, and Yoav Gilad. 2017. <span>“Batch Effects and the Effective Design of Single-Cell Gene Expression Studies.”</span> <em>Scientific Reports</em> 7 (1): 1–15.
</div>
<div id="ref-van2008visualizing" class="csl-entry">
Van der Maaten, Laurens, and Geoffrey Hinton. 2008. <span>“Visualizing Data Using t-SNE.”</span> <em>Journal of Machine Learning Research</em> 9 (11).
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Yunshun Chen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
